plugins/test/framework.h

```diff
@@ -28,10 +29,51 @@
 
 namespace service_extensions_samples {
 
+// Parameters to customize Context behaviors.
+struct ContextOptions {
+  // Wasm logging output file.
+  std::ofstream log_file;
+  // Static time returned to wasm.
+  absl::Time clock_time = absl::UnixEpoch();
+};
+
+// Buffer class to handle copying to and from an individual BODY chunk.
+class Buffer : public proxy_wasm::BufferBase {
+ public:
+  Buffer() = default;
+
+  // proxy_wasm::BufferInterface
+  size_t size() const override;
+  proxy_wasm::WasmResult copyTo(proxy_wasm::WasmBase* wasm, size_t start,
+                                size_t length, uint64_t ptr_ptr,
+                                uint64_t size_ptr) const override;
+  proxy_wasm::WasmResult copyFrom(size_t start, size_t length,
+                                  std::string_view data) override;
+
+  // proxy_wasm::BufferBase
+  void clear() override {
+    proxy_wasm::BufferBase::clear();
+    owned_string_buffer_ = "";
+  }
+  Buffer* set(std::string_view data) {
+    proxy_wasm::BufferBase::set(data);
+    return this;
+  }
+
+  std::string get() {
+    if (made_body_mutation_) {
+      return owned_string_buffer_;
+    }
+    return std::string(data_);
+  }
+
+ private:
+  bool made_body_mutation_;
+  // Buffer for a chunk post-mutation.
+  std::string owned_string_buffer_;
```

Comment:
I don't see the need to define these additional fields. The fields already inherited from BufferBase are:

```
  std::string_view data_;
  std::unique_ptr<char[]> owned_data_;
  uint32_t owned_data_size_;
```

With these, you can represent made_body_mutation_ by whether or not owned_data_ is set (i.e. non-null). So get() would be something like:

```
std::string get() {
  return owned_data_ != nullptr ? std::string(owned_data_.get(), owned_data_size_) : std::string(data_);
}
```

However, it's probably better to instead want to provide a string_view accessor, so that inspecting buffer contents doesn't require a copy:

```
std::string_view get() {
  return owned_data_ != nullptr ? std::string_view(owned_data_.get(), owned_data_size_) : data_;
}
```